util/quotapool: various features and fixes

ajwerner opened the pull request:
This code is being pulled from ongoing work on admission control but this
change seemed isolated enough to warrant a separate PR.

Release note: None

Commit:
util/quotapool: add Len() method to return the current queue length

This code is being pulled from ongoing work on admission control but this
change seemed isolated enough to warrant a separate PR.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38587)
<!-- Reviewable:end -->

Commit:
util/quotapool: add optional OnAcquire function

This change is used in upcoming work on admission control but felt sufficiently
isolated to warrant its own PR. This function is especially useful for
recording metrics after an acquisition has occurred.

Release note: None

Commit:
util/quotapool: fix name propagation

andreimatei reviewed (commented):
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38587#-:-LjIT2yv6j85AMs9pw3T:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
